File: LeetCode/python3/406.py
class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        res, people_list, people_map = [], [], {}
        for p in people:
            if p[0] not in people_map:
                people_map[p[0]] = [p[1]]
                people_list.append(p[0])
            else:
                people_map[p[0]].append(p[1])
        people_list.sort()
        for p in people_list[::-1]:
            people_map[p].sort()
            for i in range(len(people_map[p])):
                res.insert(people_map[p][i], [p, people_map[p][i]])
        return res

File: LeetCode/cpp/406.cpp
class Solution {
public:
    vector<vector<int>> reconstructQueue(vector<vector<int>> &people) {
        vector<vector<int>> res;
        unordered_map<int, vector<int>> people_map;
        vector<int> people_set;
        for (auto &p:people) {
            if (people_map.find(p[0]) == people_map.end())
                people_set.push_back(p[0]);
            people_map[p[0]].push_back(p[1]);
        }
        sort(people_set.begin(), people_set.end());
        for (auto p = people_set.rbegin(); p < people_set.rend(); ++p) {
            sort(people_map[*p].begin(), people_map[*p].end());
            for (int j = 0; j < people_map[*p].size(); ++j)
                res.insert(res.begin() + people_map[*p][j],
                           vector<int>{*p, people_map[*p][j]});
        }
        return res;
    }
};
